File: src/integrationTest/java/com/github/albertosh/flitetrakr/reactive/usecases/cheapestconnection/CheapestConnectionUseCaseTest.java
package com.github.albertosh.flitetrakr.reactive.usecases.cheapestconnection;

import com.github.albertosh.flitetrakr.model.Connection;
import com.github.albertosh.flitetrakr.reactive.services.ConnectionService;
import com.github.albertosh.flitetrakr.reactive.services.IConnectionService;

import org.junit.Before;
import org.junit.Test;

import rx.Observable;
import rx.observers.TestSubscriber;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.core.Is.is;


public class CheapestConnectionUseCaseTest {

    private IConnectionService service;
    private CheapestConnectionUseCase cheapestConnectionUseCase;

    @Before
    public void setUp() throws Exception {
        service = new ConnectionService();
        cheapestConnectionUseCase = new CheapestConnectionUseCase(service);

        Observable.merge(
                service.addConnection(Connection.fromString("NUE-FRA-43")),
                service.addConnection(Connection.fromString("NUE-AMS-67")),
                service.addConnection(Connection.fromString("FRA-AMS-17")),
                service.addConnection(Connection.fromString("FRA-LHR-27")),
                service.addConnection(Connection.fromString("LHR-NUE-23")))
                .subscribe();
    }

    @Test
    public void existingConnection() throws Exception, CheapestConnectionUseCaseError {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("NUE")
                .to("AMS")
                .build();
        TestSubscriber<CheapestConnectionUseCaseOutput> observer = TestSubscriber.create();
        cheapestConnectionUseCase.execute(input).subscribe(observer);
        observer.awaitTerminalEvent();


        // Then
        assertThat(observer.getOnNextEvents(), hasSize(1));
        CheapestConnectionUseCaseOutput result = observer.getOnNextEvents().get(0);
        assertThat(result.getCities(), contains("NUE", "FRA", "AMS"));
        assertThat(result.getPrice(), is(60));
    }

    @Test
    public void sameOriginAndDestiny() throws Exception, CheapestConnectionUseCaseError {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("LHR")
                .to("LHR")
                .build();
        TestSubscriber<CheapestConnectionUseCaseOutput> observer = TestSubscriber.create();
        cheapestConnectionUseCase.execute(input).subscribe(observer);
        observer.awaitTerminalEvent();

        // Then
        assertThat(observer.getOnNextEvents(), hasSize(1));
        CheapestConnectionUseCaseOutput result = observer.getOnNextEvents().get(0);
        assertThat(result.getCities(), contains("LHR", "NUE", "FRA", "LHR"));
        assertThat(result.getPrice(), is(93));
    }

    @Test
    public void nonExistingConnection() throws Exception {
        // Given

        // When
        CheapestConnectionUseCaseInput input = new CheapestConnectionUseCaseInput.Builder()
                .from("NUE")
                .to("DXB")
                .build();

        TestSubscriber<CheapestConnectionUseCaseOutput> observer = TestSubscriber.create();
        cheapestConnectionUseCase.execute(input).subscribe(observer);
        observer.awaitTerminalEvent();

        // Then
        assertThat(observer.getOnNextEvents(), hasSize(0));
        assertThat(observer.getOnErrorEvents(), hasSize(1));
        CheapestConnectionUseCaseError error = (CheapestConnectionUseCaseError) observer.getOnErrorEvents().get(0);
        assertThat(error.getErrorType(),
                is(CheapestConnectionUseCaseError.ErrorType.CONNECTION_NOT_FOUND));
    }

}

File: flitetrakr_sync_kotlin/src/integrationTest/kotlin/com/github/albertosh/flitetrakr/usecases/CheapestConnectionUseCaseTest.kt
package com.github.albertosh.flitetrakr.usecases

import io.kotlintest.specs.ShouldSpec

class CheapestConnectionUseCaseTest : ShouldSpec() {

    init {
        "CheapestConnectionUseCase" {

            should("throw an error if the connection doesn't exist") {
                val service = buildService()
                val usecase = CheapestConnectionUseCase(service)

                val exception: CheapestConnectionUseCaseError? =
                        try {
                            usecase.execute(
                                    CheapestConnectionUseCaseInput(
                                            from = "NUE",
                                            to = "DXB"
                                    ))
                            null
                        } catch (e: CheapestConnectionUseCaseError) {
                            e
                        }

                exception shouldBe CheapestConnectionUseCaseError.connectionNotFound
            }

            should("return the minimum price of the connection") {
                val service = buildService()
                val usecase = CheapestConnectionUseCase(service)

                val output = usecase.execute(
                        CheapestConnectionUseCaseInput(
                                from = "NUE",
                                to = "AMS"
                        ))

                output shouldBe CheapestConnectionUseCaseOutput(
                        cities = listOf("NUE", "FRA", "AMS"),
                        price = 60
                )
            }

            should("look for a connection around the globe if the origin and destination is the same") {
                val service = buildService()
                val usecase = CheapestConnectionUseCase(service)

                val output = usecase.execute(
                        CheapestConnectionUseCaseInput(
                                from = "LHR",
                                to = "LHR"
                        ))

                output shouldBe CheapestConnectionUseCaseOutput(
                        cities = listOf("LHR", "NUE", "FRA", "LHR"),
                        price = 93
                )
            }

        }
    }

}
